drivers: serial: stm32: add RTS/CTS hardware flow control

GeorgeCGV opened the pull request:
Adds RTS CTS hardware flow control.

**WIP**
- :white_check_mark: ~~missing dts arm st pinctrl rts cts pin definitions~~ (_not used_)
- :white_check_mark: add L4 ( STM32L452X) pinmux (**RTS/CTS only**)
- :white_check_mark: add F4 ( STM32F411X, STM32F446X, STM32F469X/STM32F479X (#13176)) pinmux (**RTS/CTS only**)
- :white_check_mark: reformat L4 pinmux that it looks consistent, and in order (_incr. pin#, AF#_)
- :white_check_mark: dts binding, initial value propagation
- :white_check_mark: validate with the board (_custom_)

_Restart of stagnated PR #12761_

GeorgeCGV commented:
Added F4 CTS RTS pinmux for popular STM32F411X  and STM32F446X

Commit:
drivers: pinmux: stm32: reformat L4 pinmux

Reformat L4 pinmux to be consistent with other pinmux files,
and for a better reading.

Signed-off-by: Georgij Cernysiov <g.cernysiov@elco-automation.de>

GeorgeCGV commented:
Fix incomplete L4 USART6 RX definition and separated reformated pinmux into separate commit for easing the review process.

GeorgeCGV commented:
Added dts bindings, and fixup. Tried on a board.

Commit:
drivers: serial: stm32: dts binding, and fixup for flow control

Allows to enable initial RTS/CTS hardware flow control
in the dts.

Co-authored-by: Benoit Leforestier <benoit.leforestier@gmail.com>
Signed-off-by: Georgij Cernysiov <g.cernysiov@elco-automation.de>

Commit:
drivers: pinmux: stm32: add F4 RTS/CTS definitions

Adds USART RTS/CTS definitions based on STM32F469X/STM32F479X.

Co-authored-by: Benoit Leforestier <benoit.leforestier@gmail.com>
Signed-off-by: Georgij Cernysiov <g.cernysiov@elco-automation.de>

erwango commented:
@GeorgeCGV, #12761 has been updated and is not actually staled.
Can you get along with @Benichou34 and see which one we keep?
At least for this PR it would be fair to include @Benichou34's copyright, as you mention this is a restart of his work.

GeorgeCGV commented:
@erwango I didn't modify any copyright information on any of the files, and I am not sure where this shall be added

erwango commented:
> @erwango I didn't modify any copyright information on any of the files, and I am not sure where this shall be added

My bad, I meant "sign-off-by".

GeorgeCGV commented:
@erwango not sure about multiple 'Signed-off-by', usually additional people are added with 'Co-authored-by'. Modified all commits where any work could cross with @Benichou34

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -178,6 +178,36 @@ static inline u32_t uart_stm32_get_databits(struct device *dev)
+
+#ifdef CONFIG_LPUART_1
+	if (IS_LPUART_INSTANCE(UartInstance)) {
+		LL_LPUART_SetHWFlowCtrl(UartInstance, hwctrl);
```
It appears LL LPUART and LL USART APIs are quite close and identical in the way we use them.
So we can clean up part of this and call LL_USART_SetHWFlowCtrl for all (LP)U(S)ART instances.
Btw, this is already the case in various tx/rc related functions.

Additional clean up could also be done in recent configure functions, if you're in for some code optimization.
If needed we can put a comment somewhere to state that and warn that this is still the case with new LL API calls that may be introduced.

GeorgeCGV replied:
I think such clean-up (touching all LPUART calls) would be better as a separate PR.
Reworked HW flow control related part, and added comments there for the future.

erwango commented on drivers/pinmux/stm32/pinmux_stm32l4x.h:
```diff
@@ -133,7 +133,7 @@
 	(STM32_PINMUX_ALT_FUNC_8 | STM32_PUSHPULL_PULLUP)
 #define STM32L4X_PINMUX_FUNC_PC7_USART6_RX  \
-	(STM32_PINMUX_ALT_FUNC_8 | STM32_PUPDR_NO_PUL
+	(STM32_PINMUX_ALT_FUNC_8 | STM32_PUPDR_NO_PULL)
```
I wonder how it is possible this hasn't been seen before.
Can you push this as a separate bug fix commit so we can get it in 1.14?

GeorgeCGV replied:
PR #14054

erwango commented on dts/bindings/serial/st,stm32-lpuart.yaml:
```diff
@@ -29,4 +29,10 @@ properties:
+    hw-flow-control:
+      type: boolean
+      category: optional
+      description: use RTS/CTS flow control
```
This might be misleading to people not used to dts (even if already used in other bindings), this is not a boolean that expect a true/false value. Either it is present and means enabled or it is not there andthen flow control is disabled. "Boolean" applies to the generated  flag, not to the node property.
What about: when set, RTC/CTS flow control is enable at boot time.

GeorgeCGV replied:
Yep, fits

erwango commented on drivers/pinmux/stm32/pinmux_stm32f4.h:
```diff
@@ -418,6 +418,8 @@
 #define STM32F4_PINMUX_FUNC_PG9_USART6_RX   \
 	(STM32_PINMUX_ALT_FUNC_8 | STM32_PUSHPULL_NOPULL)
 
+#define STM32F4_PINMUX_FUNC_PG8_USART6_RTS \
```
Are these STM32F479X specific?
If yes, they should be under appropriate flags, else rewrite the commit message/title

GeorgeCGV replied:
Not really, reworded.

erwango commented on soc/arm/st_stm32/stm32f4/dts_fixup.h:
```diff
@@ -116,6 +117,7 @@
 #define DT_USART_2_IRQ				DT_ST_STM32_USART_40004400_IRQ_0
 #define DT_UART_STM32_USART_2_CLOCK_BITS	DT_ST_STM32_USART_40004400_CLOCK_BITS
 #define DT_UART_STM32_USART_2_CLOCK_BUS	DT_ST_STM32_USART_40004400_CLOCK_BUS
+#define DT_UART_STM32_USART_2_HW_FLOW_CONTROL	DT_ST_STM32_USART_40004400_HW_FLOW_CONTROL
```
DT_ST_STM32_**UART**_40004400_HW_FLOW_CONTROL (cf #14555)

GeorgeCGV replied:
Rebased

Commit:
drivers: serial: stm32: add serial hw flow control

Adds RTS CTS hardware flow control support.

Signed-off-by: Georgij Cernysiov <g.cernysiov@elco-automation.de>

Commit:
drivers: pinmux: stm32: add F4 RTS/CTS definitions

Adds common F4 RTS/CTS definitions.

Co-authored-by: Benoit Leforestier <benoit.leforestier@gmail.com>
Signed-off-by: Georgij Cernysiov <g.cernysiov@elco-automation.de>

Commit:
drivers: pinmux: stm32: add L4 RTS/CTS definitions

Adds common L4 RTS/CTS definitions.

Signed-off-by: Georgij Cernysiov <g.cernysiov@elco-automation.de>

GeorgeCGV commented on drivers/serial/uart_stm32.c:
```diff
@@ -178,6 +178,28 @@ static inline u32_t uart_stm32_get_databits(struct device *dev)
+
+	/* LPUART and U(S)ART have the same base and use the same
+	 * register and masks for hw flow control,
+	 * validate for newly added SoC
```
Moved to the main file comment as a `@note`.

GeorgeCGV commented on drivers/pinmux/stm32/pinmux_stm32f4.h:
```diff
@@ -21,6 +21,8 @@
 /* Port A */
 #define STM32F4_PINMUX_FUNC_PA0_PWM2_CH1    \
 	(STM32_PINMUX_ALT_FUNC_1 | STM32_PUSHPULL_PULLUP)
+#define STM32F4_PINMUX_FUNC_PA0_USART2_CTS \
```
Did this, and corrected the commit messages related to the pinmux to point to the pinmux instead of the serial driver.

GeorgeCGV commented:
Rebased, and corrected `uart_stm32_cfg2ll_hwctrl `, `uart_config_flow_control` to not violate MISRA C:2012, 16.3; and added comments for both functions.
